builtin/readline_osh.py

```diff
@@ -49,46 +48,67 @@ def __exit__(self, type, value, traceback):
 class BindXCallback(object):
     """A callable we pass to readline for executing shell commands."""
 
-    def __init__(self, eval):
-        # type: (Eval) -> None
+    def __init__(self, eval, mem, errfmt):
+        # type: (Eval, Mem, ui.ErrorFormatter) -> None
         self.eval = eval
+        self.mem = mem
+        self.errfmt = errfmt
 
     def __call__(self, cmd, line_buffer, point):
-        # type: (str, str, int) -> Tuple[int, str, str]
+        # type: (str, str, int) -> Tuple[int, str, int]
         """Execute a shell command through the evaluator.
 
         Args:
           cmd: The shell command to execute
           line_buffer: The current line buffer
           point: The current cursor position
         """
-        print("Setting READLINE_LINE to: %s" % line_buffer)
-        print("Setting READLINE_POINT to: %s" % point)
-        print("Executing cmd: %s" % cmd)
 
-        # TODO: add READLINE_* env vars later
-        # assert isinstance(line_buffer, str)
-        # self.mem.SetNamed(location.LName("READLINE_LINE"),
-        #                  value.Str(line_buffer),
-        #                  scope_e.GlobalOnly,
-        #                  flags=SetExport)
+        # Set READLINE_* env vars so they're available in case the command uses them
+        state.ExportGlobalString(self.mem, "READLINE_LINE", line_buffer)
+        state.ExportGlobalString(self.mem, "READLINE_POINT", str(point))
 
         # TODO: refactor out shared code from Eval, cache parse tree?
 
         cmd_val = cmd_eval.MakeBuiltinArgv([cmd])
         status = self.eval.Run(cmd_val)
 
-        return (status, line_buffer, str(point))
+        # Retrieve READLINE_* env vars and compare for changes
+        readline_line = self._get_rl_env_var('READLINE_LINE')
+        readline_point = self._get_rl_env_var('READLINE_POINT')
+        post_line_buffer = readline_line if readline_line is not None else line_buffer
+        post_point = int(
+            readline_point) if readline_point is not None else point
+
+        self.mem.Unset(location.LName('READLINE_LINE'), scope_e.GlobalOnly)
+        self.mem.Unset(location.LName('READLINE_POINT'), scope_e.GlobalOnly)
+
+        return (status, post_line_buffer, post_point)
+
+    def _get_rl_env_var(self, envvar_name):
+        # type: (str) -> Optional[str]
+        """Retrieve the value of an env var, return None if undefined"""
+
+        envvar_val = self.mem.GetValue(envvar_name, scope_e.GlobalOnly)
+        if envvar_val.tag() == value_e.Str:
```

Comment:
To test if this value is either `Str` or `Undef` or else, please use this `with tagswitch` idiom

https://oils.pub/release/latest/doc/oils-repo/mycpp/README.html#with-tagstr_switch-switch-statement

You can also grep the code for `tagswitch` -- we use it all over

It translates to a switch statement in C++!